### PR TITLE
Travis: run itests in parallel, get ~60% speedup overall

### PR DESCRIPTION
**Pull request by guggero**
Depends on https://github.com/lightningnetwork/lnd/pull/4268, only the last 8 commits are new.

This PR splits all integration tests in 4 distinct groups and runs them in parallel. This creates a bit of overhead by creating the test harness (mining btc node, chain backend btc node and two lnd nodes) four times.
But that easily offset by running the tests in parallel and saturating the CPU much better.

### Before
(master at 9c9ce15ca679b15b73d00c6e912ea93d1458b5eb)
itest_master_before_speedup

### After
(this PR at 15420639767e0bffe01d8fa22cc73c5de13390da)
itest_4655_after_speedup

### Speedup
 - Btcd: 62%
 - Bitcoind with txindex: 58%
 - Bitcoind without txindex: 59%
 - Neutrino: 64%
 - Btcd ARM: 70%
 - Btcd Windows: 55%

### Flakes

There still are some flakes in the itests. But I feel that they are on an acceptable level with all the recent fixes. Still, the following list gives an indication of the still open flakes and the ones that were fixed by this PR or #4721.

#### `TestLightningNetworkDaemon/group4-9of18/bitcoind/immediate_payment_after_channel_opened`
```
*errors.errorString payment failed: FAILED
.../github.com/lightningnetwork/lnd/lntest/itest/lnd_test.go:13995 (0xea45da)
     sendAndAssertSuccess: t.Fatalf("payment failed: %v", result.Status)
.../github.com/lightningnetwork/lnd/lntest/itest/lnd_test.go:1584 (0xe57f85)
     testPaymentFollowingChannelOpen: sendAndAssertSuccess(t, net.Alice, &routerrpc.SendPaymentRequest{
```
- Fixed 1714394a :heavy_check_mark: 

#### `TestLightningNetworkDaemon/group3-17of18/btcd/async_bidirectional_payments`
```
test_harness.go:88: Failed: (async bidirectional payments): exited with error: 
*errors.errorString unable to close channel: error while waiting for broadcast tx: tx not seen before context timeout
.../github.com/lightningnetwork/lnd/lntest/itest/lnd_test.go:176 (0xe19947)
```
- Found problem, connection issue between miner and backend btcd node :hourglass: 

#### `TestLightningNetworkDaemon/group3-18of18/neutrino/test_multi-hop_htlc/committype=legacy/receiver_chain_claim`
```
test_harness.go:91: Error outside of test: *errors.errorString unable to find Carol's force close tx in mempool: wanted 2, found 1 txs in mempool: [62dea1ffd7a9408616d809e171d18de03c9f55351297e8cf750a83fd63168596]
.../github.com/lightningnetwork/lnd/lntest/itest/lnd_test.go:9411 (0xe8251e)
     assertDLPExecuted: t.Fatalf("unable to find Carol's force close tx in mempool: %v",
```
 - Fixed 0f2827aa :heavy_check_mark: 

#### `TestLightningNetworkDaemon/group2-8of15/bitcoind/abandonchannel`
```
test_harness.go:88: Failed: (abandonchannel): exited with error: 
*errors.errorString channel shouldn't be found in the channel graph!
.../github.com/lightningnetwork/lnd/lntest/itest/lnd_test.go:13106 (0xe9dfe5)
     testAbandonChannel: t.Fatalf("channel shouldn't be found in the channel " +
.../github.com/lightningnetwork/lnd/lntest/itest/test_harness.go:112 (0xe300ee)
     (*harnessTest).RunTestCase: testCase.test(h.lndHarness, h)
.../github.com/lightningnetwork/lnd/lntest/itest/lnd_test.go:14208 (0xec7448)
     TestLightningNetworkDaemon.func4: ht.RunTestCase(testCase)
/home/travis/.gimme/versions/go1.15.3.linux.amd64/src/testing/testing.go:1123 (0x51cd0f)
     tRunner: fn(t)
/home/travis/.gimme/versions/go1.15.3.linux.amd64/src/runtime/asm_amd64.s:1374 (0x46f721)
     goexit: BYTE	$0x90	// NOP
```

#### `TestLightningNetworkDaemon/group2-5of15/neutrino/revoked_uncooperative_close_retribution`
```
test_harness.go:88: Failed: (revoked uncooperative close retribution): exited with error: 
*errors.errorString error while waiting for channel close: rpc error: code = DeadlineExceeded desc = context deadline exceeded
.../github.com/lightningnetwork/lnd/lntest/itest/lnd_test.go:8221 (0xe7b89e)
     testRevokedCloseRetribution: t.Fatalf("error while waiting for channel close: %v", err)
.../github.com/lightningnetwork/lnd/lntest/itest/test_harness.go:112 (0xe2d32e)
     (*harnessTest).RunTestCase: testCase.test(h.lndHarness, h)
.../github.com/lightningnetwork/lnd/lntest/itest/lnd_test.go:14208 (0xec4628)
     TestLightningNetworkDaemon.func4: ht.RunTestCase(testCase)
/home/travis/.gimme/versions/go1.15.3.linux.amd64/src/testing/testing.go:1123 (0x51bd6f)
     tRunner: fn(t)
/home/travis/.gimme/versions/go1.15.3.linux.amd64/src/runtime/asm_amd64.s:1374 (0x46f721)
     goexit: BYTE	$0x90	// NOP
```

### CODE REVIEW

File: lntest/itest/lnd_test_list_group1_test.go [hunk 1]
```diff
@@ -0,0 +1,78 @@
+// +build rpctest
+
+package itest
+
+var testCasesGroup1 = []*testCase{
```
Comment (bhandras) [inline lntest/itest/lnd_test_list_group1_test.go:5]: We could split programmatically instead, so we don't have to have four files for the groups. Also would make it easier to parametrize how parallel we want to run (I realize this is a bit more difficult to do but with predefined group names it'd be viable)

File: lntest/node.go [hunk 2]
```diff
@@ -104,6 +109,15 @@ func nextAvailablePort() int {
 	panic("no ports available for listening")
 }
 
+// ApplyPortOffset reads the passed -portoffset flag and applies it to the next
+// available port.
+func ApplyPortOffset() {
```
Comment (bhandras) [inline lntest/node.go:118]: We could use the less error prone `getFreePort()` function from `channeldb/kvdb/etcd/embed.go` in which case we won't need the `lastPort` and `portOffset`.

File: lntest/itest/lnd_channel_backup_test.go [hunk 3]
```diff
@@ -1012,6 +1012,10 @@ func testChanRestoreScenario(t *harnessTest, net *lntest.NetworkHarness,
 	require.Contains(t.t, err.Error(), "cannot close channel with state: ")
 	require.Contains(t.t, err.Error(), "ChanStatusRestored")
 
+	// Increase the fee estimate so that the following force close tx will
+	// be cpfp'ed in case of anchor commitments.
+	net.SetFeeEstimate(30000)
```
Comment (bhandras) [inline lntest/itest/lnd_channel_backup_test.go:1017]: Wow, encountered this flake with etcd so many times... Thanks :+1:

**Review by bhandras — COMMENTED** [review]
Awesome! My comments are just suggestions, otherwise looks good!

File: lntest/itest/lnd_test_list_group1_test.go (hunk 1 above)
Comment (guggero) [reply lntest/itest/lnd_test_list_group1_test.go]: The reason why I started with the manual split was because it seemed that some tests were more flaky depending on the order they were executed (as if there was some left-over state from previous runs). But it does make more sense to do the split programmatically and fix the flakes instead.

File: lntest/node.go (hunk 2 above)
Comment (guggero) [reply lntest/node.go]: I started using that but now I'm not sure if it actually is better. Because it does a `listener.Close()` just before returning the port number, that port now essentially is free again. And in a system where things happen in parallel, this could lead to conflicts. I'm going to revert back to the `lastPort` variable but make the offset more dynamic.

Comment (bhandras) [reply lntest/node.go]: One way to help mitigate this is that we could use port ranges where slices try and look for open ports.

Comment (guggero) [reply lntest/node.go]: Isn't that exactly what the `nextAvailablePort()` in combination with the offset does?

File: lntest/itest/lnd_test.go [hunk 4]
```diff
@@ -53,6 +53,60 @@ import (
...
+	defaultRunTranche uint = 0
+)
+
+var (
+	// testCasesSplitParts is the number of tranches the test cases should
+	// be split into. By default this is set to 1, so no splitting happens.
+	// If this value is increased, then the -runtranche flag must be
+	// specified as well to indicate which part should be run in the current
+	// invocation.
+	testCasesSplitTranches = flag.Uint(
```
Comment (bhandras) [inline lntest/itest/lnd_test.go:72]: suggestion: would the name of `numParallelSplits` work better?

File: lntest/itest/lnd_test.go [hunk 5]
```diff
@@ -14098,10 +14152,16 @@ func getPaymentResult(stream routerrpc.Router_SendPaymentV2Client) (
...
-	if len(testsCases) == 0 {
+	if len(allTestCases) == 0 {
 		t.Skip("integration tests not selected with flag 'rpctest'")
 	}
 
+	// Parse testing flags that influence our test execution.
+	logDir := lntest.GetLogDir()
+	require.NoError(t, os.MkdirAll(logDir, 0700))
+	testCases, trancheIndex, trancheOffset := getTestCaseSplitTranche()
+	lntest.ApplyPortOffset(uint32(trancheIndex) * 100)
```
Comment (bhandras) [inline lntest/itest/lnd_test.go:14163]: If I understand correctly this means we're offsetting by 100  for ports. Is that enough?

File: lntest/itest/lnd_test.go [hunk 6]
```diff
@@ -53,6 +54,60 @@ import (
...
+	testCasesRunTranche = flag.Uint(
+		"runtranche", defaultRunTranche, "run the tranche of the "+
+			"split test cases with the given (0-based) index",
+	)
+)
+
+// getTestCaseSplitTranche returns the sub slice of the test cases that should
+// be run as the current split tranche as well as the index and slice offset of
+// the tranche.
+func getTestCaseSplitTranche() ([]*testCase, uint, uint) {
```
Comment (bhandras) [inline lntest/itest/lnd_test.go:89]: :+1:

**Review by bhandras — APPROVED** [review]
LGTM, much nicer to have the programmatic splits! :1st_place_medal:

File: lntest/itest/lnd_test.go (hunk 5 above)
Comment (guggero) [reply lntest/itest/lnd_test.go]: Increased it to 1000.

File: lntest/itest/lnd_test_list_on_test.go [hunk 7]
```diff
@@ -3,6 +3,10 @@
 package itest
 
 var allTestCases = []*testCase{
+	{
+		name: "test multi-hop htlc",
+		test: testMultiHopHtlcClaims,
```
Comment (wpaulino) [inline lntest/itest/lnd_test_list_on_test.go:8]: What's the rationale for this change?

**Review by wpaulino — APPROVED** [review]

File: lntest/itest/lnd_test.go (hunk 4 above)
Comment (guggero) [reply lntest/itest/lnd_test.go:72]: Yeah, maybe. But I'd rather not rename everything again :sweat_smile:

File: lntest/itest/lnd_test_list_on_test.go (hunk 7 above)
Comment (guggero) [reply lntest/itest/lnd_test_list_on_test.go:8]: Added a comment to the commit:

> To make sure the test that takes the longest overall time is always
started first, independent of the number of test tranches we run, we
move it to the beginning of the list. Because that test involves a lot
of waiting, it allows us to play around with the number of tranches more
efficiently.
